documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md

```diff
@@ -0,0 +1,610 @@
+- Utility files for helper functions for connecting to Elasticsearch and
+  Postgres (e.g.
+  [`es_helpers.py`](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/es_helpers.py))
+
+The
+[`Dockerfile`](https://github.com/WordPress/openverse/blob/main/ingestion_server/Dockerfile)
+can be copied from the existing ingestion server. It should be updated to
+reference the new file structure, and to expose only a single port, which should
+be distinguished from the ports currently in use by the ingestion server (8001
+and 8002). Other necessary files, including `env.docker`, `.dockerignore`,
+`Pipfile`, and `gunicorn.conf.py` can all be copied in from the existing
+ingestion server as well.
+
+Finally we will update the monorepo's root
+[`docker-compose.yml`](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/docker-compose.yml)
+to add a new `catalog-indexer-worker` service. Its build context should point to
+the nested `data_refresh/indexer_worker` directory, and it should map the
+exposed port to enable the API to be reached by the catalog.
+
+When this work is complete, it should be possible to run `just catalog/shell`
+and curl the new indexer worker. The existing ingestion-server and
+indexer-worker services are unaffected (it is still possible to run legacy data
+refreshes locally and in production).
+
+### Implement distributed reindexing locally
+
+In this step we will add tasks to the data refresh DAGs to orchestrate the
+distributed reindex. At the end of this step, it will be possible to run a
+distributed reindex _locally_, but because the infrastructure work to create the
+ASGs is not complete, it can not be run on production yet. The following code
+can all be refactored from
+[`distributed_reindex_scheduler.py`](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/distributed_reindex_scheduler.py).
+
+- Use
+  [`describe_auto_scaling_groups`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/autoscaling/client/describe_auto_scaling_groups.html)
+  and filter by tags to select the appropriate ASG for the desired environment.
+  (Skips in local env.)
+- Use
+  [`set_desired_capacity`](https://boto3.amazonaws.com/v1/documentation/api/1.26.86/reference/services/autoscaling/client/set_desired_capacity.html)
+  to increase the desired capacity of the ASG to the desired number of workers,
+  depending on the environment. This will cause the ASG to begin spinning up
+  instances. (Skips in local env.)
+- Use
+  [`describe_auto_scaling_groups`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/autoscaling/client/describe_auto_scaling_groups.html)
+  to poll the ASG until all instances have been started, and get the EC2
+  instance IDs. (Skips in local env.)
+- Use dynamic task mapping to generate a Sensor for each of the instance IDs,
+  and ping its `healthcheck` endpoint until it passes.
```

Comment:
Good call. I was thinking about pinging the healthcheck for the api, rather than the instance itself -- but I had not considered that it's possible to define custom health checks, which looks like it should be possible. That would make this possible.

I was thinking about the local healthcheck endpoint and I think it's okay to not have that, since locally we'll have the container running when the stack comes up (rather than starting/stopping it to try to simulate the asg or anything like that).